# Análisis ConnectaTel

Como **analista de datos**, tu objetivo es evaluar el **comportamiento de los clientes** de una empresa de telecomunicaciones en Latinoamérica, ConnectaTel. 

Trabajaremos con información registrada **hasta el año 2024**, lo cual permitirá analizar el comportamiento del negocio dentro de ese periodo.

Para ello trabajarás con tres datasets:  

- **plans.csv** → información de los planes actuales (precio, minutos incluidos, GB incluidos, costo por extra)  
- **users.csv** → información de los clientes (edad, ciudad, fecha de registro, plan, churn)  
- **usage.csv** → detalle del **uso real** de los servicios (llamadas y mensajes)  

Deberás **explorar**, **limpiar** y **analizar** estos datos para construir un **perfil estadístico** de los clientes, detectar **comportamientos atípicos** y crear **segmentos de clientes**.  

Este análisis te permitirá **identificar patrones de consumo**, **diseñar estrategias de retención** y **sugerir mejoras en los planes** ofrecidos por la empresa.

> 💡 Antes de empezar, recuerda pensar de forma **programática**: ¿qué pasos necesitas? ¿En qué orden? ¿Qué quieres medir y por qué?


--- 
## 🧩 Paso 1: Cargar y explorar

Antes de limpiar o combinar los datos, es necesario **familiarizarte con la estructura de los tres datasets**.  
En esta etapa, validarás que los archivos se carguen correctamente, conocerás sus columnas y tipos de datos, y detectarás posibles inconsistencias.

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:**  
Tener los **3 datasets listos en memoria**, entender su contenido y realizar una revisión preliminar.

**Instrucciones:**  
- Importa las librerías necesarias (por ejemplo `pandas`, `seaborn`, `matplotlib.pyplot`)
- Carga los archivos CSV usando `pd.read_csv()`:
  - **`/datasets/plans.csv`**  
  - **`/datasets/users_latam.csv`**  
  - **`/datasets/usage.csv`**  
- Guarda los DataFrames en las variables: `plans`, `users`, `usage`.  
- Muestra las primeras filas de cada DataFrame usando `.head()`.


In [70]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [71]:

plans = pd.read_csv('/datasets/plans.csv')
users = pd.read_csv('/datasets/users_latam.csv')
usage = pd.read_csv('/datasets/usage.csv')

In [72]:
plans.head()

,plan_name,messages_included,gb_per_month,minutes_included,usd_monthly_pay,usd_per_gb,usd_per_message,usd_per_minute
0,Basico,100,5,100,12,1.2,0.08,0.10
1,Premium,500,20,600,25,1.0,0.05,0.07


In [73]:
users.head()

,user_id,first_name,last_name,age,city,reg_date,plan,churn_date
0,10000,Carlos,Garcia,38,Medellín,2022-01-01 00:00:00.000000000,Basico,NaN
1,10001,Mateo,Torres,53,?,2022-01-01 06:34:17.914478619,Basico,NaN
2,10002,Sofia,Ramirez,57,CDMX,2022-01-01 13:08:35.828957239,Basico,NaN
3,10003,Mateo,Ramirez,69,Bogotá,2022-01-01 19:42:53.743435858,Premium,NaN
4,10004,Mateo,Torres,63,GDL,2022-01-02 02:17:11.657914478,Basico,NaN


In [74]:
usage.head()

,id,user_id,type,date,duration,length
0,1,10332,call,2024-01-01 00:00:00.000000000,0.09,NaN
1,2,11458,text,2024-01-01 00:06:30.969774244,NaN,39.0
2,3,11777,text,2024-01-01 00:13:01.939548488,NaN,36.0
3,4,10682,call,2024-01-01 00:19:32.909322733,1.53,NaN
4,5,12742,call,2024-01-01 00:26:03.879096977,4.84,NaN


**Tip:** Si no usas `print()` la tabla se vera mejor.

### 1.2 Exploración de la estructura de los datasets

**🎯 Objetivo:**  
Conocer la **estructura de cada dataset**, revisar cuántas filas y columnas tienen, identificar los **tipos de datos** de cada columna y detectar posibles **inconsistencias o valores nulos** antes de iniciar el análisis.

**Instrucciones:**  
- Revisa el **número de filas y columnas** de cada dataset usando `.shape`.  
- Usa `.info()` en cada DataFrame para obtener un **resumen completo** de columnas, tipos de datos y valores no nulos.  

In [75]:

# revisar el número de filas y columnas de cada dataset
print("plans", plans.shape)
print("users", users.shape)
print("usage", usage.shape)

plans (2, 8)
users (4000, 8)
usage (40000, 6)


In [76]:
# inspección de plans con .info()
plans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   plan_name          2 non-null      object 
 1   messages_included  2 non-null      int64  
 2   gb_per_month       2 non-null      int64  
 3   minutes_included   2 non-null      int64  
 4   usd_monthly_pay    2 non-null      int64  
 5   usd_per_gb         2 non-null      float64
 6   usd_per_message    2 non-null      float64
 7   usd_per_minute     2 non-null      float64
dtypes: float64(3), int64(4), object(1)
memory usage: 256.0+ bytes


In [77]:
# inspección de users con .info()
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user_id     4000 non-null   int64 
 1   first_name  4000 non-null   object
 2   last_name   4000 non-null   object
 3   age         4000 non-null   int64 
 4   city        3531 non-null   object
 5   reg_date    4000 non-null   object
 6   plan        4000 non-null   object
 7   churn_date  466 non-null    object
dtypes: int64(2), object(6)
memory usage: 250.1+ KB


In [78]:
# inspección de usage con .info()
usage.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        40000 non-null  int64  
 1   user_id   40000 non-null  int64  
 2   type      40000 non-null  object 
 3   date      39950 non-null  object 
 4   duration  17924 non-null  float64
 5   length    22104 non-null  float64
dtypes: float64(2), int64(2), object(2)
memory usage: 1.8+ MB


---

## 🧩Paso 2: Identificación de problemas de calidad de datos

### 2.1 Revisión de valores nulos

**🎯 Objetivo:**  
Detectar la presencia y magnitud de valores faltantes para evaluar si afectan el análisis o requieren imputación/eliminación.

**Instrucciones:**  
- Cuenta valores nulos por columna para cada dataset.
- Calcula la proporción de nulos por columna para cada dataset.

El dataset `plans` solamente tiene 2 renglones y se puede observar que no tiene ausentes, por ello no necesita exploración adicional.

<br>
<details>
<summary>Haz clic para ver la pista</summary>
Usa `.isna().sum()` para contar valores nulos y usa `.isna().mean()` para calcular la proporción.

In [79]:
# cantidad de nulos para users
print(users.isnull().sum())
print(users.isnull().mean() * 100)

user_id          0
first_name       0
last_name        0
age              0
city           469
reg_date         0
plan             0
churn_date    3534
dtype: int64
user_id        0.000
first_name     0.000
last_name      0.000
age            0.000
city          11.725
reg_date       0.000
plan           0.000
churn_date    88.350
dtype: float64


In [80]:
# cantidad de nulos para usage
print(usage.isnull().sum())
print(usage.isnull().mean() * 100)

id              0
user_id         0
type            0
date           50
duration    22076
length      17896
dtype: int64
id           0.000
user_id      0.000
type         0.000
date         0.125
duration    55.190
length      44.740
dtype: float64


✍️ **Comentario**: Haz doble clic en este bloque y escribe tu diagnóstico al final del bloque. Incluye qué ves y que acción recomendarías para cada caso.


type: 0.000 (0%) → sin nulos, no requiere acción
	•	date: 0.125 (12.5%) → está en el rango de 5-30%, así que conviene investigar antes de decidir
	•	duration: 55.190 (55.2%) → alto, pero no llega al 80-90%, así que también amerita investigar (aunque con un porcentaje tan alto, imputar puede ser riesgoso)
	•	length: 44.740 (44.7%) → similar caso, investigar

Un borrador de diagnóstico que podrías adaptar y pegar en el bloque de comentario:

Valores nulos: Las columnas date (12.5%), duration (55.2%) y length (44.7%) presentan valores faltantes; type no tiene nulos. Dado que duration y length superan el 30%, se investigará su origen antes de decidir entre imputar (ej. con la mediana) o dejar como nulos, ya que eliminarlas podría perder mucha información. Para date, al estar en el rango medio, se evaluará el patrón de ausencia (aleatorio vs. sistemático) antes de imputar

### 2.2 Detección de valores inválidos y sentinels

🎯 **Objetivo:**  
Identificar sentinels: valores que no deberían estar en el dataset.

**Instrucciones:**
- Explora las columnas numéricas con **un resumen estadístico** y describe brevemente que encontraste.
- Explora las columnas categóricas **relevantes**, revisando sus valores únicos y describe brevemente que encontraste.


El dataset `plans` solamente tiene 2 renglones, por ello no necesita exploración adicional.

In [81]:
# Contar cuántas filas tienen age = -999 (antes de limpiar)
cantidad_999 = (users['age'] == -999).sum()
porcentaje_999 = (cantidad_999 / len(users)) * 100
print(f"Filas con age = -999: {cantidad_999} ({porcentaje_999:.1f}%)")

Filas con age = -999: 55 (1.4%)


In [82]:
# explorar columnas numéricas de users
users.describe()

,user_id,age
count,4000.000000,4000.000000
mean,11999.500000,33.739750
std,1154.844867,123.232257
min,10000.000000,-999.000000
25%,10999.750000,32.000000
50%,11999.500000,47.000000
75%,12999.250000,63.000000
max,13999.000000,79.000000


La columna user_id va de 10,000 a 13,999, sin valores fuera de rango. Es un identificador único y consistente; no requiere ninguna acción.
* La columna age presenta un valor mínimo de -999, un sentinel imposible para una edad real. Se recomienda reemplazarlo por NaN y evaluar si se imputa con la mediana o se deja como nulo.

In [83]:
# explorar columnas numéricas de usage
usage.describe()

,id,user_id,duration,length
count,40000.00000,40000.000000,17924.000000,22104.000000
mean,20000.50000,12002.405975,5.202237,52.127398
std,11547.14972,1157.279564,6.842701,56.611183
min,1.00000,10000.000000,0.000000,0.000000
25%,10000.75000,10996.000000,1.437500,37.000000
50%,20000.50000,12013.000000,3.500000,50.000000
75%,30000.25000,13005.000000,6.990000,64.000000
max,40000.00000,13999.000000,120.000000,1490.000000


•	Las columnas id y user_id van en rangos consistentes y sin valores negativos ni fuera de lo esperado. Son indicadores, no requieren ninguna acción.
	•	Las columnas duration y length presentan un alto porcentaje de valores faltantes, más del 50% en ambas, por lo que se investigará si están relacionados con la columna type (MAR) antes de decidir si se imputan o se dejan como nulos. Además, length muestra un valor máximo de 1490, muy por encima del percentil 75.

In [84]:

# explorar columnas categóricas de users
columnas_users = ['city', 'plan'] 

- La columna `type` contiene dos categorías: “call” (llamada) y “text” (mensaje), sin valores inválidos. Es justamente la columna clave que explica por qué duration y length tienen tantos nulos: duration solo aplica a registros tipo “call” y length solo a tipo “text”, confirmando que son valores MAR (Missing At Random) relacionados con type.


---
✍️ **Comentario**: Haz doble clic en este bloque y escribe tu diagnóstico. Incluye qué ves y que acción recomendarías para cada caso. 

**Valores inválidos o sentinels**  
- ¿En qué columnas encontraste valores inválidos o sentinels?
- En las columnas age (dataset users) y length (dataset usage):

	•	age: valor mínimo de -999, imposible para una edad real (el resto de los datos van de 18 a 79 años). Es un sentinel clásico usado para marcar “edad desconocida”.
	•	length: máximo de 1490, muy alejado del percentil 75 (64) y de la media (52). Es un outlier que sugiere error de captura, no necesariamente un sentinel, pero igual amerita revisión.

duration y user_id/id no mostraron valores fuera de rango evidentes en el describe().
- ¿Qué acción tomarías?
- 	Reemplazar -999 por NaN (no es un valor real, es un sentinel).
	•	Luego decidir: como es solo un valor puntual, se puede imputar con la mediana de age o dejarlo como nulo si no es crítico para el análisis.

length (outlier de 1490 en usage)

	•	No lo eliminaría de inmediato solo por ser alto — primero graficaría un boxplot o histograma para ver cuántos registros están muy por encima del rango normal.
	•	Si son pocos casos aislados y no tienen explicación lógica (ej. no corresponden a llamadas largas legítimas), los trataría como outliers y los recortaría (capping) o los excluiría del análisis estadístico, dejando nota de por qué.
	•	Si son varios y siguen un patrón (ej. siempre en cierto plan o ciudad), podría ser un dato real de uso intensivo, no un error — en ese caso los dejaría.

### 2.3 Revisión y estandarización de fechas

**🎯 Objetivo:**  
Asegurar que las columnas de fecha estén correctamente formateadas y detectar años fuera de rango que indiquen errores de captura.

**Instrucciones:**  
- Convierte las columnas de fecha a tipo fecha y asegurate de que el código sea a prueba de errores.  
- Revisa cuántas veces aparece cada año.
- Identifica fechas imposibles (ej. años futuros o negativos).

Toma en cuenta que tenemos datos registrados hasta el año 2024.

In [85]:

# Convertir a fecha la columna reg_date de users
users['reg_date'] = pd.to_datetime(users['reg_date'], errors='coerce')


In [86]:
# Convertir a fecha la columna date de usage
usage['date'] = pd.to_datetime(usage['date'], errors='coerce')

In [87]:
# Revisar los años presentes en reg_date de users
users['reg_date'].dt.year.value_counts().sort_index()

2022    1314
2023    1316
2024    1330
2026      40
Name: reg_date, dtype: int64

En `reg_date`, se identificaron 40 registros con año 2026, lo cual es imposible ya que los datos solo deben llegar hasta 2024. Estas fechas se consideran errores de captura; se recomienda tratarlas como valores inválidos (reemplazar por NaT) y evaluar si se pueden corregir con otra fuente o dejarlas como nulas..

In [88]:
# Revisar los años presentes en date de usage
usage['date'].dt.year.value_counts().sort_index()


2024.0    39950
Name: date, dtype: int64

Los nulos en duration y length son MAR (Missing At Random), ya que dependen directamente de la columna type: duration está prácticamente completo solo cuando type es “call” (99.9% nulo en “text”), y length está completo solo cuando type es “text” (99.9% nulo en “call”). Al confirmarse esta dependencia, se decide dejar los nulos como están (NaN), ya que no aplican para ese tipo de registro y no tiene sentido imputarlos. Basaremos el análisis en estas fechas.

Diagnóstico: Se detectaron 40 registros en reg_date con año 2026, una fecha imposible ya que los datos están registrados solo hasta 2024. Esto sugiere un error de captura (por ejemplo, un dígito mal ingresado al guardar la fecha).
Acción recomendada: Marcar estos registros como valores nulos (NaT) usando errors='coerce' en la conversión, o filtrarlos del análisis para no distorsionar las métricas basadas en fechas de registro. Si es posible, contactar al equipo de origen de datos para corregir la causa del error en la fuente

---

## 🧩Paso 3: Limpieza básica de datos

### 3.1 Corregir sentinels y fechas imposibles
**🎯 Objetivo:**  
Aplicar reglas de limpieza para reemplazar valores sentinels y corregir fechas imposibles.

**Instrucciones:**  
- En `age`, reemplaza el sentinel **-999** con la mediana.
- En `city`, reemplaza el sentinel `"?"` por valores nulos (`pd.NA`).  
- Marca como nulas (`pd.NA`) las fechas fuera de rango.

In [89]:
# Reemplazar -999 por la mediana de age
age_mediana = users.loc[users['age'] != -999, 'age'].median()
users['age'] = users['age'].replace(-999, age_mediana)

# Verificar cambios
users['age'].describe()

count    4000.000000
mean       48.136000
std        17.689919
min        18.000000
25%        33.000000
50%        48.000000
75%        63.000000
max        79.000000
Name: age, dtype: float64

In [90]:
# Reemplazar ? por NA en city
users['city'] = users['city'].replace('?', pd.NA)

# Verificar cambios
users['city'].value_counts(dropna=False)

Bogotá      808
CDMX        730
Medellín    616
NaN         565
GDL         450
Cali        424
MTY         407
Name: city, dtype: int64

In [91]:

# Marcar fechas futuras como NA para reg_date
users['reg_date'] = pd.to_datetime(users['reg_date'], errors='coerce')  # re-asegura el tipo fecha
users['reg_date'] = users['reg_date'].where(users['reg_date'].dt.year <= 2024, pd.NaT)

# Verificar cambios
users['reg_date'].dt.year.value_counts(dropna=False).sort_index()

2022.0    1314
2023.0    1316
2024.0    1330
NaN         40
Name: reg_date, dtype: int64

### 3.2 Corregir sentinels y fechas imposibles
**🎯 Objetivo:**  
Decidir qué hacer con los valores nulos según su proporción y relevancia.

**Instrucciones:**
- Verifica si los nulos en `duration` y `length` son **MAR**(Missing At Random) revisando si dependen de la columna `type`.  
  Si confirmas que son MAR, **déjalos como nulos** y justifica la decisión.

In [92]:
# Verificación MAR en usage (Missing At Random) para duration
usage.groupby('type')['duration'].apply(lambda x: x.isnull().mean() * 100)


type
call     0.000000
text    99.927576
Name: duration, dtype: float64

In [93]:
# Verificación MAR en usage (Missing At Random) para length
usage.groupby('type')['length'].apply(lambda x: x.isnull().mean() * 100)


type
call    99.932991
text     0.000000
Name: length, dtype: float64

In [94]:
# explorar columnas categóricas de users
columnas_user = ['city', 'plan']

for col in columnas_user:
    print(f"--- {col} ---")
    print(users[col].value_counts())
    print()

--- city ---
Bogotá      808
CDMX        730
Medellín    616
GDL         450
Cali        424
MTY         407
Name: city, dtype: int64

--- plan ---
Basico     2595
Premium    1405
Name: plan, dtype: int64



Haz doble clic aquíy escribe que tu diagnostico de nulos en `duration` y `length`son MAR (Missing At Random), ya que dependen directamente de la columna type: duration es 0% nulo cuando type es “call” y 99.9% nulo cuando es “text”; length es 99.9% nulo cuando type es “call” y 0% cuando es “text”. Esto confirma que cada columna solo aplica a su tipo de registro correspondiente. Se decide dejar los nulos como están (NaN), ya que no corresponde imputarlos: no tiene sentido inventar una duración para un mensaje de texto ni una longitud para una llamada.


---

## 🧩Paso 4: Summary statistics de uso por usuario


### 4.1 Agrupación por comportamiento de uso

🎯**Objetivo**: Resumir las variables clave de la tabla `usage` **por usuario**, creando métricas que representen su comportamiento real de uso histórico. 

**Instrucciones:**: 
1. Construye una tabla agregada de `usage` por `user_id` que incluya:
- número total de mensajes  
- número total de llamadas  
- total de minutos de llamadas

2. Renombra las columnas para que tengan nombres claros:  
- `cant_mensajes`  
- `cant_llamadas`  
- `cant_minutos_llamada`
3. Combina esta tabla con `users`.

In [95]:

# Columnas auxiliares
usage["is_text"] = (usage["type"] == "text").astype(int) #conocer el total de mensajes
usage["is_call"] = (usage["type"] == "call").astype(int) #conocer el total de llamadas
usage["call_minutes"] = usage["duration"] * usage["is_call"] #conocer los minutos de llamada

In [96]:
# Agrupar información por usuario
usage_agg = usage.groupby("user_id")[["is_text", "is_call", "call_minutes"]].sum().reset_index()

# observar resultado
usage_agg.head(3)

,user_id,is_text,is_call,call_minutes
0,10000,7,3,23.70
1,10001,5,10,33.18
2,10002,5,2,10.74


In [97]:
user_profile = users.merge(usage_agg, on='user_id', how='left')

In [98]:
# Rellenar con 0 a los usuarios sin actividad registrada en usage
user_profile[['cant_mensajes', 'cant_llamadas', 'cant_minutos_llamada']] = \
    user_profile[['cant_mensajes', 'cant_llamadas', 'cant_minutos_llamada']].fillna(0)

KeyError: "None of [Index(['cant_mensajes', 'cant_llamadas', 'cant_minutos_llamada'], dtype='object')] are in the [columns]"

### 4.2 Resumen estadístico por usuario durante el 2024

🎯 **Objetivo:** Analizar las columnas numéricas y categóricas de los usuarios, para identificar rangos, valores extremos y distribución de los datos antes de continuar con el análisis.

**Instrucciones:**  
1. Para las columnas **numéricas** relevantes, obtén un resumen estadístico (media, mediana, mínimo, máximo, etc.).  
2. Para la columna **categórica** `plan`, revisa la distribución en **porcentajes** de cada categoría.

In [ ]:
# Renombrar columnas
usage_agg = usage_agg.rename(columns={
    "is_text": "cant_mensajes",
    "is_call": "cant_llamadas",
    "call_minutes": "cant_minutos_llamada"
})

# observar resultado
usage_agg.head(3)

In [ ]:
# Combinar la tabla agregada con el dataset de usuarios
user_profile = users.merge(usage_agg, on='user_id', how='left')
user_profile.head(5)

In [ ]:
# Resumen estadístico de las columnas numéricas
user_profile[['age', 'cant_mensajes', 'cant_llamadas', 'cant_minutos_llamada']].describe()


---

## 🧩Paso 5: Visualización de distribuciones (uso y clientes) y outliers


### 5.1 Visualización de Distribuciones

🎯 **Objetivo:**  
Entender visualmente cómo se comportan las variables clave tanto de **uso** como de **clientes**, observar si existen diferencias según el tipo de plan, y analizar la **forma de la distribución**.

**Instrucciones:**  
Graficar **histogramas** para las siguientes columnas:  
- `age` (edad de los usuarios)
- `cant_mensajes`
- `cant_llamadas`
- `total_minutos_llamada` 

Después de cada gráfico, escribe un **insight** respecto al plan y la variable, por ejemplo:  
- "Dentro del plan Premium, hay mayor proporción de..."  
- "Los usuarios Básico tienden a hacer ... llamadas y enviar ... mensajes."  o "No existe algún patrón."
- ¿Qué tipo de distribución tiene ? (simétrica, sesgada a la derecha o a la izquierda) 

**Hint**  
Para cada histograma, 
- Usa `hue='plan'` para ver cómo varían las distribuciones según el plan (Básico o Premium).
- Usa `palette=['skyblue','green']`
- Agrega título y etiquetas

In [ ]:

# Histograma para visualizar la edad (age)
sns.histplot(data=user_profile, x='age', hue='plan', palette=['skyblue', 'green'])
plt.title('Distribución de Edad por Plan')
plt.xlabel('Edad')
plt.ylabel('Cantidad')
plt.show()

💡Insights: 
- Distribución de edad por plan: La distribución de age es relativamente uniforme entre los 18 y 80 años para ambos planes, sin un sesgo claro hacia usuarios jóvenes o mayores. En todos los rangos de edad, el plan Básico mantiene una proporción consistentemente mayor que Premium, lo cual es coherente con la distribución general (64.9% Básico vs. 35.1% Premium) vista anteriormente. No se observan grupos etarios con preferencia marcada por un plan específico.

In [ ]:

# Histograma para visualizar la cant_mensajes
sns.histplot(data=user_profile, x='cant_mensajes', hue='plan', palette=['skyblue', 'green'])
plt.title('Distribución de Cantidad de Mensajes por Plan')
plt.xlabel('Cantidad de Mensajes')
plt.ylabel('Cantidad de Usuarios')
plt.show()


In [ ]:

# Histograma para visualizar la cant_llamadas
sns.histplot(data=user_profile, x='cant_llamadas', hue='plan', palette=['skyblue', 'green'])
plt.title('Distribución de Cantidad de Llamadas por Plan')
plt.xlabel('Cantidad de Llamadas')
plt.ylabel('Cantidad de Usuarios')
plt.show()


💡Insights:
- Distribución de cant_mensajes por plan: distribución unimodal sesgada a la derecha, con pico entre 4 y 6 mensajes y una cola larga hasta 17. El plan Básico se mantiene consistentemente por encima de Premium en cada barra, en proporción similar a la distribución general (65% Básico vs 35% Premium), sin un patrón de uso distintivo entre planes.
- Distribución de cant_llamadas por plan: distribución unimodal sesgada a la derecha, con pico entre 3 y 5 llamadas y una cola hasta 15. Igual que en cant_mensajes, Básico se mantiene proporcionalmente por encima de Premium en todo el rango, sin diferencias marcadas de comportamiento entre planes.

In [ ]:

# Histograma para visualizar la cant_minutos_llamada
sns.histplot(data=user_profile, x='cant_minutos_llamada', hue='plan', palette=['skyblue', 'green'])
plt.title('Distribución de Minutos de Llamada por Plan')
plt.xlabel('Minutos de Llamada')
plt.ylabel('Cantidad de Usuarios')
plt.show()

 Distribución de minutos de llamada por plan: La distribución está fuertemente sesgada a la derecha, concentrada entre 0 y 40 minutos para ambos planes, con Básico (azul) siempre por encima de Premium (verde) en cantidad de usuarios. Se confirma visualmente la presencia de outliers: una pequeña cola de usuarios dispersos entre 100 y 160 minutos, muy alejados del grueso de la distribución, coincidiendo con los outliers detectados anteriormente mediante el método IQR (límite superior ≈ 61 minutos).
 

### 5.2 Identificación de Outliers

🎯 **Objetivo:**  
Detectar valores extremos en las variables clave de **uso** y **clientes** que podrían afectar el análisis, y decidir si requieren limpieza o revisión adicional.

**Instrucciones:**  
- Usa **boxplots** para identificar visualmente outliers en las siguientes columnas:  
  - `age` 
  - `cant_mensajes`
  - `cant_llamadas`
  - `total_minutos_llamada`  
- Crea un **for** para generar los 4 boxplots automáticamente.
<br>

- Después de crear los gráfico, responde si **existen o no outliers** en las variables.  
- Si hay outliers, crea otro bucle para calcular los límites de esas columnas usando el **método IQR** y decide qué hacer con ellos.
  - Si solamente hay outliers de un solo lado, no es necesario calcular ambos límites.

**Hint:**
- Dentro del bucle, usa `plt.title(f'Boxplot: {col}')` para que el título cambie acorde a la columna.

In [ ]:

# Visualizando usando BoxPlot
columnas_numericas = ['age', 'cant_mensajes', 'cant_llamadas', 'cant_minutos_llamada']

for col in columnas_numericas:
    sns.boxplot(x=user_profile[col], color='skyblue')
    plt.title(f'Boxplot: {col}')
    plt.show()

💡Insights: 
- Age no presenta outliers evidentes
- cant_mensajes presenta alguns outliers pero son valores plausibles de usuario muy activos
- cant_llamadas presenta outliers 
- cant_minutos_llamada presenta numerosos outliers

In [ ]:

# Calcular límites con el método IQR
columnas_limites = []


for col in columnas_numericas:
    Q1 = user_profile[col].quantile(0.25)
    Q3 = user_profile[col].quantile(0.75)
    IQR = Q3 - Q1
    limite_inferior = Q1 - 1.5 * IQR
    limite_superior = Q3 + 1.5 * IQR
    print(f'{col}: límite inferior = {limite_inferior:.2f}, límite superior = {limite_superior:.2f}')
    columnas_limites.append(col)

In [ ]:
# Revisa los límites superiores y el max, para tomar la decisión de mantener los outliers o no
user_profile[columnas_limites].describe()

💡Insights:
- cant_mensajes: se mantienen los outliers. Los valores hasta 17 son posibles para usuarios de alto consumo.
- cant_llamadas: se mantienen los outliers. Los valores hasta 15 llamadas son razonables.
- cant_minutos_llamada: el máximo excede considerablemente el rango esperado. Se investigará el registro puntual y, de no justificarse, se aplicará capping en vez de eliminación.

---

## 🧩Paso 6: Segmentación de Clientes

### 6.1 Segmentación de Clientes Por Uso

🎯 **Objetivo:** Clasificar a cada usuario en un grupo de uso (Bajo uso, Uso medio, Alto uso) basándose en la cantidad de llamadas y mensajes registrados.

**Instrucciones:**  
- Crea una nueva columna llamada `grupo_uso` en el dataframe `user_profile`.
- Usa comparaciones lógicas (<, >) para evaluar las condiciones de llamadas y mensajes y asigna:
  - `'Bajo uso'` cuando llamadas < 5 y mensajes < 5
  - `'Uso medio'` cuando llamadas < 10 y mensajes < 10
  - `'Alto uso'` para el resto de casos

In [ ]:


# Crear columna grupo_uso
def clasificar_uso(row):
    llamadas = row['cant_llamadas']
    mensajes = row['cant_mensajes']
    
    if llamadas < 5 and mensajes < 5:
        return 'Bajo uso'
    elif llamadas < 10 and mensajes < 10:
        return 'Uso medio'
    else:
        return 'Alto uso'


In [ ]:
# verificar cambios
user_profile['grupo_uso'] = user_profile.apply(clasificar_uso, axis=1)

In [ ]:
user_profile[['cant_llamadas', 'cant_mensajes', 'grupo_uso']].head()

La mayoría de los clientes se concentra en el grupo de Uso medio, representando aproximadamente el 74% de la base total (2,943 de 4,000 usuarios). Esto sugiere que la mayoría de los clientes de ConnectaTel tienen un consumo moderado y estable, lo cual representa una oportunidad para diseñar un plan intermedio ajustado a este comportamiento, en lugar de forzarlos a elegir entre Básico o Premium

### 6.2 Segmentación de Clientes Por Edad

🎯 **Objetivo:**: Clasificar a cada usuario en un grupo por **edad**.

**Instrucciones:**  
- Crea una nueva columna llamada `grupo_edad` en el dataframe `user_profile`.
- Usa comparaciones lógicas (<, >) para evaluar las condiciones y asigna:
  - `'Joven'` cuando age < 30
  - `'Adulto'` cuando age < 60
  - `'Adulto Mayor'` para el resto de casos

In [ ]:

# Definir la función de clasificación por edad
def clasificar_edad(row):
    age = row['age']
    if age < 30:
        return 'Joven'
    elif age < 60:
        return 'Adulto'
    else:
        return 'Adulto Mayor'

# Crear la columna grupo_edad
user_profile['grupo_edad'] = user_profile.apply(clasificar_edad, axis=1)


### 6.3 Visualización de la Segmentación de Clientes

🎯 **Objetivo:** Visualizar la distribución de los usuarios según los grupos creados: **grupo_uso** y **grupo_edad**.

**Instrucciones:**  
- Crea dos gráficos para las variables categóricas `grupo_uso` y `grupo_edad`.
- Agrega título y etiquetas a los ejes en cada gráfico.

In [ ]:
# Visualización de los segmentos por uso
sns.countplot(data=user_profile, x='grupo_uso', color='skyblue')
plt.title('Distribución de Usuarios por Grupo de Uso')
plt.xlabel('Grupo de Uso')
plt.ylabel('Cantidad de Usuarios')
plt.show()

In [ ]:
# Visualización de los segmentos por edad
sns.countplot(data=user_profile, x='grupo_edad', color='lightgreen')
plt.title('Distribución de Usuarios por Grupo de Edad')
plt.xlabel('Grupo de Edad')
plt.ylabel('Cantidad de Usuarios')
plt.show()


---
## 🧩Paso 7: Insight Ejecutivo para Stakeholders

🎯 **Objetivo:** Traducir los hallazgos del análisis en conclusiones accionables para el negocio, enfocadas en segmentación, patrones de uso y oportunidades comerciales.

**Preguntas a responder:** 
- ¿Qué problemas tenían originalmemte los datos?

- Se detectaron varios problemas de calidad en los datos:

	•	La columna age contenía el valor sentinel -999 en un grupo de registros (valor imposible para una edad real), corregido reemplazándolo por la mediana.
	•	La columna city usaba el símbolo ”?” como marcador de dato faltante en 469 de 4000 registros (11.7%), reemplazado por valores nulos para tratarlos correctamente.
	•	En reg_date se encontraron 40 registros con año 2026 (fecha imposible, ya que los datos llegan solo hasta 2024), marcados como nulos.
	•	En usage, las columnas duration y length tenían nulos que resultaron ser MAR (dependientes del tipo de registro: mensajes sin duración de llamada, y llamadas sin longitud de mensaje), por lo que se dejaron como nulos justificadamente.


¿Qué porcentaje, o cantidad de filas, de esa columna representaban?

age: el valor sentinel -999 apareció en 55 registros, equivalente a un 1.4% del total de 4000 filas.
age: 55 registros (1.4%)
	•	city: 469 registros (11.7%)
	•	reg_date: 40 registros (1%)

- ¿Qué segmentos de clientes identificaste y cómo se comportan según su edad y nivel de uso?

- En los tres grupos de edad, “Uso medio” es la categoría dominante — la mayoría de tus clientes, sin importar su edad, tienen un consumo moderado.
	•	Los Adultos son el grupo con más usuarios en Alto uso (154 de 279 en total, ~55% de todos los usuarios de alto uso), y también el grupo más numeroso en general (2018 usuarios).
	•	En términos relativos, si sacas el porcentaje dentro de cada grupo de edad: Adulto tiene ~7.6% en Alto uso, Adulto Mayor ~6.1%, y Joven ~6.7% — están bastante parejos, pero Adulto tiene una ligera ventaja.
Los tres grupos de edad muestran una distribución similar, concentrándose mayoritariamente en Uso medio. Sin embargo, el segmento Adulto no solo es el más numeroso (2018 usuarios, ~50% de la base), sino que también concentra la mayor cantidad absoluta de usuarios de Alto uso (154), consolidándose como el segmento más valioso tanto en volumen como en potencial de ingresos premium.
- ¿Qué segmentos parecen más valiosos para ConnectaTel y por qué?

- Los usuarios de Alto uso, independientemente de su edad, representan el mayor valor potencial: generan más ingresos por consumo de llamadas y mensajes, y son candidatos naturales para upselling hacia planes premium. El grupo_uso con Alto uso es particularmente atractivo por combinar capacidad de gasto con consumo elevado.

  
- ¿Qué patrones de uso extremo (outliers) encontraste y qué implican para el negocio?

- Al analizar cant_mensajes, cant_llamadas y cant_minutos_llamada con boxplots y el método IQR, se identificaron outliers superiores. Estos valores extremos podrían representar: (a) clientes de uso corporativo o intensivo mal clasificados en planes personales, o posibles errores de registro que ameritan revisión.
Al aplicar el método IQR sobre las variables de consumo, se identificaron 46 outliers en cant_mensajes, 30 en cant_llamadas y 96 en cant_minutos_llamada — siendo esta última la que concentra el mayor número de usuarios con consumo atípicamente alto. Esto sugiere que existe un grupo de clientes (principalmente ligado a duración de llamadas) cuyo uso supera ampliamente lo esperado para un plan estándar, lo cual podría indicar clientes con necesidades de tipo corporativo mal clasificados en planes personales, o una oportunidad clara de upselling hacia planes premium.

- ¿Qué recomendaciones harías para mejorar la oferta actual de planes o crear nuevos planes basados en los segmentos y patrones detectados?

- Crear un plan intermedio dirigido a usuarios de Uso medio, que actualmente podrían estar mal ajustados entre el plan Básico y el Premium.
	•	Diseñar una oferta específica para los outliers de alto consumo, ofreciéndoles un plan de “uso ilimitado” o corporativo, capturando ingresos que hoy podrían estarse perdiendo o generando fricción.
	•	Enfocar campañas de retención en el segmento Adulto Mayor con Bajo uso, revisando si presentan mayor tasa de baja (churn_date).
	•	Corregir en el origen los procesos de captura de age y reg_date para evitar que se repitan estos errores en futuras cargas de datos.

✍️ **Escribe aquí tu análisis ejecutivo:**

### Análisis ejecutivo

⚠️ **Problemas detectados en los datos**
La columna age contenía el valor sentinel -999 en varios registros, un valor imposible que fue reemplazado por la mediana para no distorsionar el análisis.
	•	La columna city tenía el símbolo ”?” como marcador de dato faltante, reemplazado por valores nulos (NA) para tratarlo correctamente como ausencia de información.
	•	En reg_date se detectaron 40 registros con año 2026, una fecha imposible dado que los datos están registrados solo hasta 2024; estos se marcaron como nulos (NaT), indicando un probable error de captura.
	•	En usage, los nulos en duration y length resultaron ser MAR (Missing At Random), ya que dependen del tipo de registro (type): los mensajes de texto no tienen duración de llamada, y las llamadas no tienen longitud de mensaje, por lo que se dejaron como nulos de forma justificada.

🔍 **Segmentos por Edad**
-Se clasificó a los usuarios en Bajo uso, Uso medio y Alto uso, según su cantidad de llamadas y mensajes.

Esto sugiere que Los clientes de Alto uso podrían representar una oportunidad de upselling hacia planes premium o corporativos, ya que su consumo elevado de llamadas y mensajes indica capacidad de gasto y necesidades que el plan Básico probablemente no cubre del todo


💡 **Recomendaciones**
Diseñar un plan intermedio para los usuarios de Uso medio que actualmente podrían estar sobrepagando o subutilizando su plan actual.
	•	Dirigir campañas de retención hacia el segmento de Adulto Mayor con Bajo uso, si se detecta que tienden a dar de baja el servicio (churn_date).
	•	Corregir en el origen los errores de captura de edad y fecha de registro para evitar que se repitan en futuras cargas de datos.

---

## 🧩Paso 8 Cargar tu notebook y README a GitHub

🎯 **Objetivo:**  
Entregar tu análisis de forma **profesional**, **documentada** y **versionada**, asegurando que cualquier persona pueda revisar, ejecutar y entender tu trabajo.



### Opción A : Subir archivos desde la interfaz de GitHub (UI)

1. Descarga este notebook (`File → Download .ipynb`).  
2. Entra a tu repositorio en GitHub (por ejemplo `telecom-analysis` o `sprint7-final-project`).  
3. Sube tu notebook **Add file → Upload files**.  

---

### Opción B : Guardar directo desde Google Colab

1. Abre tu notebook en Colab.  
2. Ve a **File → Save a copy in GitHub**.  
3. Selecciona el repositorio y la carpeta correcta (ej: `notebooks/`).  
4. Escribe un mensaje de commit claro, por ejemplo:  
    - `feat: add final ConnectaTel analysis`
    - `agregar version final: Análisis ConnectaTel`
5. Verifica en GitHub que el archivo quedó en el lugar correcto y que el historial de commits se mantenga limpio.

---

Agrega un archivo `README.md` que describa de forma clara:
- el objetivo del proyecto,  
- los datasets utilizados,  
- las etapas del análisis realizadas,  
- cómo ejecutar el notebook (por ejemplo, abrirlo en Google Colab),  
- una breve guía de reproducción.
---

Link a repositorio público del proyecto: `LINK a tu repo aquí`

In [ ]:

# Link a repositorio público del proyecto:


In [ ]:
# Tabla cruzada: edad vs uso
pd.crosstab(user_profile['grupo_edad'], user_profile['grupo_uso'])

In [ ]:
# Cuántos usuarios son outliers en cada columna
for col in columnas_limites:
    Q1 = user_profile[col].quantile(0.25)
    Q3 = user_profile[col].quantile(0.75)
    IQR = Q3 - Q1
    limite_inferior = Q1 - 1.5 * IQR
    limite_superior = Q3 + 1.5 * IQR
    
    outliers = user_profile[(user_profile[col] < limite_inferior) | (user_profile[col] > limite_superior)]
    print(f"{col}: {len(outliers)} outliers")